### PR TITLE
4021: Preventing Save Game Loading Player Name NPE

### LINKED ISSUE #4021: NPE while loading a save-file
### Environment
Linux Megamek 0.49.10, java 11.0.13

### Description
Loading a save game fails. After some digging through the logs and the error message it seems to be unhappy about some Arrow IV attacks made by infantry.

### Files
[night.sav.gz](https://github.com/MegaMek/megamek/files/10149910/night.sav.gz)

Log  message bellow;
19:35:13,427 INFO  [megamek.MegaMek] {main}
megamek.MegaMek.initializeLogging(MegaMek.java:115) - Starting MegaMek v0.49.10
	Build Date: N/A
	Today: 2022-12-04
	Origin Project: MegaMek
	Java Vendor: Ubuntu
	Java Version: 11.0.13
	Platform: Linux 5.11.0-49-generic (amd64)
	System Locale: en_US
	Total memory available to MegaMek: 2 GB

19:35:13,429 INFO  [megamek.MegaMek] {main}
megamek.MegaMek.handleLegacyLogging(MegaMek.java:128) - Redirecting System.out, System.err, and Throwable::printStackTrace output to legacy.log

19:35:13,546 INFO  [megamek.client.ui.preferences.SuitePreferences] {main}
megamek.client.ui.preferences.SuitePreferences.loadFromFile(SuitePreferences.java:118) - Loading user preferences from: mmconf/mm.preferences

19:35:13,583 INFO  [megamek.client.ui.preferences.SuitePreferences] {main}
megamek.client.ui.preferences.SuitePreferences.loadFromFile(SuitePreferences.java:138) - Finished loading user preferences

19:35:13,815 INFO  [megamek.MegaMek] {main}
megamek.MegaMek.startGUI(MegaMek.java:315) - Starting MegaMekGUI.

19:35:21,528 ERROR [megamek.client.ui.swing.MegaMekGUI] {AWT-EventQueue-0}
megamek.client.ui.swing.MegaMekGUI.loadGame(MegaMekGUI.java:548) - Unable to load file: /home/dirk/apps/mekhq-0.49.10/savegames/night.sav.gz
com.thoughtworks.xstream.converters.ConversionException: Failed calling method
---- Debugging information ----
message             : Failed calling method
cause-exception     : java.lang.NullPointerException
cause-message       : null
method              : megamek.common.weapons.ArtilleryWeaponIndirectFireHandler.readObject()
class               : megamek.common.weapons.ArtilleryWeaponIndirectFireHandler
required-type       : megamek.common.weapons.ArtilleryWeaponIndirectFireHandler
converter-type      : com.thoughtworks.xstream.converters.reflection.SerializableConverter
line number         : 545411
class[1]            : java.util.Vector
required-type[1]    : java.util.Vector
converter-type[1]   : com.thoughtworks.xstream.converters.collections.CollectionConverter
class[2]            : megamek.common.Game
required-type[2]    : megamek.common.Game
converter-type[2]   : com.thoughtworks.xstream.converters.reflection.ReflectionConverter
version             : 1.4.14
-------------------------------
	at com.thoughtworks.xstream.core.util.SerializationMembers.callReadObject(SerializationMembers.java:136)
	at com.thoughtworks.xstream.converters.reflection.SerializableConverter.doUnmarshal(SerializableConverter.java:443)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:277)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at com.thoughtworks.xstream.converters.collections.AbstractCollectionConverter.readBareItem(AbstractCollectionConverter.java:132)
	at com.thoughtworks.xstream.converters.collections.AbstractCollectionConverter.readItem(AbstractCollectionConverter.java:117)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.addCurrentElementToCollection(CollectionConverter.java:98)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.populateCollection(CollectionConverter.java:91)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.populateCollection(CollectionConverter.java:85)
	at com.thoughtworks.xstream.converters.collections.CollectionConverter.unmarshal(CollectionConverter.java:80)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshallField(AbstractReflectionConverter.java:499)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:425)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:277)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.start(TreeUnmarshaller.java:134)
	at com.thoughtworks.xstream.core.AbstractTreeMarshallingStrategy.unmarshal(AbstractTreeMarshallingStrategy.java:32)
	at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1404)
	at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1383)
	at com.thoughtworks.xstream.XStream.fromXML(XStream.java:1277)
	at megamek.client.ui.swing.MegaMekGUI.loadGame(MegaMekGUI.java:546)
	at megamek.client.ui.swing.MegaMekGUI.lambda$new$0(MegaMekGUI.java:935)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:279)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6635)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
	at megamek.client.ui.swing.widget.MegamekButton.processMouseEvent(MegamekButton.java:242)
	at java.desktop/java.awt.Component.processEvent(Component.java:6400)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5011)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4843)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4843)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.lang.NullPointerException
	at megamek.common.weapons.WeaponHandler.readObject(WeaponHandler.java:435)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.thoughtworks.xstream.core.util.SerializationMembers.callReadObject(SerializationMembers.java:132)
	... 65 more

### LINKED ISSUE #4201: [0.49.11] Asymmetrical LOS exists between units in Double-Blind with Diagramming LOS on, when no LOS should exist between the units
### Environment
megamek.MegaMek.initializeLogging(MegaMek.java:115) - Starting MegaMek v0.49.11
	Build Date: 2022-12-22T10:48:03.564757800
	Today: 2023-02-22
	Origin Project: MegaMek
	Java Vendor: Eclipse Adoptium
	Java Version: 11.0.14
	Platform: Windows 10 10.0 (amd64)
	System Locale: en_US
	Total memory available to MegaMek: 2 GB

### Description
In rare situations, units marked as "Hidden" to the owner are visible to enemy units, but not vice versa. All LOS tools show there is no LOS between the units, and there should be no LOS between the units using any version of the game rules.
Double-blind and Diagramming LOS is on.

In the attached screenshot and save file, the Vulcan in hex 1404 is visible to the enemy team, even though 3 or more light woods exist between it and the enemy units.
No enemy units are visible to the team the Vulcan is on.

This happened the other way around in an earlier game. The Lancelot and Galleon were visible, but not the other units. So I was able to reproduce the general problem, but not the specifics.

I've checked my understanding of the standard and diagramming rules on TO pg. 80, and am pretty sure no LOS should exist between the units at all, much less asymmetrical LOS. And all within-game tools say there should be no LOS. The ruler tool says blocked by woods, LOS tool says blocked by woods, and the hex is greyed out indicating no LOS.

In this case no units were able to fire at each other in the weapons phase. I'm not sure if that was true the first time this happened.
That means this doesn't break the weapon attack phase, but it still breaks double-blind where a unit could be revealed that shouldn't be, and the owner might not be aware of that.

In the attached save file, after completing the turn by taking no actions, at the beginning of the movement phase the situation was reversed and the Vulcan could see the Lancelot and Galleon, but was not seen in return.

I have not tried it with Diagramming LOS off, but suspect that's involved.
[asymmetricalLOSusingdiagrammingLOSrules.sav.gz](https://github.com/MegaMek/megamek/files/10805737/asymmetricalLOSusingdiagrammingLOSrules.sav.gz)
asymmtricalLOS

### PR DESCRIPTION
**Pull request by Windchild292**
This fixes #4021. The first commit is cleanup while investigating the issue, the second commit is the actual fix.

I used manual XML parsing as we only need a few pieces to load in, and this prevents weird loading errors. It also now parses the version correctly if the file cannot load, which improves that handling too and prevents some other file load issues.

### CODE REVIEW

File: megamek/src/megamek/client/ui/swing/MegaMekGUI.java [hunk 1]
```diff
@@ -577,6 +599,86 @@ public String getDescription() {
...
+    private void parsePlayerNames(final Node n, final Vector<String> playerNames) {
+        if (!n.hasChildNodes()) {
+            return;
+        }
+
+        final NodeList nl = n.getChildNodes();
+        for (int i = 0; i < nl.getLength(); i++) {
+            final Node n2 = nl.item(i);
+            if ((n2.getNodeType() != Node.ELEMENT_NODE) || !n2.hasChildNodes()
+                    || !"megamek.common.Player".equals(n2.getNodeName())) {
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/MegaMekGUI.java:670]: Would it make sense to use Player.class.getName() here? For the (unlikely) event of renaming or moving stuff

File: megamek/src/megamek/client/ui/swing/MegaMekGUI.java [hunk 2]
```diff
@@ -577,6 +599,86 @@ public String getDescription() {
...
+
+    private boolean validateSaveVersion(final Node n) {
+        if (!n.hasChildNodes()) {
+            final String message = String.format(
+                    Messages.getString("MegaMek.LoadGameMissingVersion.message"),
+                    MMConstants.VERSION);
+            JOptionPane.showMessageDialog(frame, message,
+                    Messages.getString("MegaMek.LoadGameAlert.title"),
+                    JOptionPane.ERROR_MESSAGE);
+            LogManager.getLogger().error(message);
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/MegaMekGUI.java:616]: Is this missing a return false; ? or intentional fall-through?

**Review by SJuliez — APPROVED** [review]
OK as far as I can say; I have two questions or comments.

File: megamek/src/megamek/client/ui/swing/MegaMekGUI.java (hunk 1 above)
Comment (Windchild292) [reply megamek/src/megamek/client/ui/swing/MegaMekGUI.java:670]: Yes

File: megamek/src/megamek/client/ui/swing/MegaMekGUI.java (hunk 2 above)
Comment (Windchild292) [reply megamek/src/megamek/client/ui/swing/MegaMekGUI.java:616]: Dumb bug, fixed

**Pull request merged**